shelley/chain-and-ledger/formal-spec/update.tex

```diff
@@ -0,0 +1,396 @@
+  \label{fig:ts-types:pp-update}
+\end{figure}
+
+\begin{figure}[htb]
+  \begin{equation}\label{eq:pp-update-Empty}
+    \inference[PP-Update-Empty]
+    {
+      \var{pup} = \emptyset
+    }
+    {
+      \begin{array}{l}
+        \var{slot}\\
+        \var{dms}\\
+      \end{array}
+      \vdash \var{pup_s}\trans{ppup}{pup}\var{pup_s}
+    }
+  \end{equation}
+
+  \nextdef
+
+  \begin{equation}\label{eq:update-nonempty}
+    \inference[PP-Update-Nonempty]
+    {
+      \var{pup}\neq\emptyset
+      &
+      \dom{pup}\subseteq\dom{dms}
+      \\
+      \var{slot} < \firstSlot{((\epoch{slot}) + 1) - \SlotsPrior}
+    }
+    {
+      \begin{array}{l}
+        \var{slot}\\
+        \var{dms}\\
+      \end{array}
+      \vdash
+      \var{pup_s}
+      \trans{ppup}{pup}
+      \varUpdate{pup_s\unionoverrideRight pup}
+    }
+  \end{equation}
+
+  \caption{Protocol Parameter Update inference rules}
+  \label{fig:rules:pp-update}
+\end{figure}
+
+\clearpage
+
+Figure~\ref{fig:funcs:helper-updates} gives some helper functions for the
+application version update transition.
+The function $\fun{votedValue_T}$ returns
+the consensus value of update proposals in the event that at least five
+genesis keys agree.
+This function will also be used for the protocol parameters later in Section~\ref{sec:epoch}.
+Note that $\type{T}$ is an arbitrary type.
+
+%%
+%% Figure - Helper Function for Consensus of Update Proposals
+%%
+\begin{figure}[htb]
+  \begin{align*}
+      & \fun{votedValue_T} \in (\VKeyGen\mapsto\type{T}) \to \type{T}^?\\
+      & \fun{votedValue_T}~\var{vs} =
+        \begin{cases}
+          t & \exists t\in\range{vs}(|vs\restrictrange t|\geq 5) \\
+          \Nothing & \text{otherwise} \\
+        \end{cases}
+  \end{align*}
+
+  \begin{align*}
+      & \fun{newAVs} \in \Applications \to (\Slot\mapsto\Applications) \to \Applications \\
+      & \fun{newAVs}~\var{avs}~\var{favs} =
+        \begin{cases}
+          \var{avs}' & \var{favs}\neq\emptyset
+                       ,~s_m\mapsto\var{avs'}\in\var{favs}
+                       ,~s_m=\max(\dom{\var{favs}})\\
+          \var{avs} & \text{otherwise}
+        \end{cases}
+  \end{align*}
+
+  \caption{Epoch Helper Functions}
+  \label{fig:funcs:helper-updates}
+\end{figure}
+
+
+
+\begin{figure}[htb]
+  \emph{Application Version Update environment}
+  \begin{equation*}
+    \AVUpdateEnv =
+    \left(
+      \begin{array}{r@{~\in~}lr}
+        \var{slot} & \Slot & \text{current slot}\\
+        \var{dms} & \VKeyGen\mapsto\VKey & \text{genesis key delegations} \\
+      \end{array}
+    \right)
+  \end{equation*}
+  %
+  \emph{Application Version Update states}
+  \begin{equation*}
+    \AVUpdateState =
+    \left(
+      \begin{array}{r@{~\in~}lr}
+        \var{aup} & \AVUpdate & \text{application versions proposals} \\
+        \var{favs} & \Slot\mapsto\Applications & \text{future application versions} \\
+        \var{avs} & \Applications & \text{current application versions} \\
```

Comment:
In the Byron spec @dcoutts pointed out to me that we needed to keep the slot in which the application version changed. We might need the same here.

Reply:
I think `versions` is correct, since there are multiple `AppName`s

Reply:
@dnadales I made this issue so that we don't forget: #575

Reply:
Great! Thanks!